go/nbs/dynamo_table_reader.go

```diff
@@ -38,98 +35,71 @@ func (t tableNotInDynamoErr) Error() string {
 	return fmt.Sprintf("NBS table %s not present in DynamoDB table %s", t.nbs, t.dynamo)
 }
 
-func newDynamoTableReader(ddb ddbsvc, table string, h addr, chunkCount uint32, data []byte, indexCache *indexCache, tc *sizecache.SizeCache) chunkSource {
-	d.PanicIfTrue(table == "")
-	source := &dynamoTableReader{ddb: ddb, table: table, h: h, tc: tc}
-
-	var index tableIndex
-	found := false
-	if indexCache != nil {
-		indexCache.lockEntry(h)
-		defer indexCache.unlockEntry(h)
-		index, found = indexCache.get(h)
-	}
-
-	if !found {
-		index = parseTableIndex(data)
-		if indexCache != nil {
-			indexCache.put(h, index)
-		}
-	}
-
-	source.tableReader = newTableReader(index, source, fileBlockSize)
-	d.PanicIfFalse(chunkCount == source.count())
-	return source
-}
-
-func (dtr *dynamoTableReader) hash() addr {
-	return dtr.h
-}
-
-func (dtr *dynamoTableReader) ReadAtWithStats(p []byte, off int64, stats *Stats) (n int, err error) {
-	t1 := time.Now()
-	if data, present := dynamoTableCacheMaybeGet(dtr.tc, dtr.hash()); present {
-		defer func() {
-			stats.MemBytesPerRead.Sample(uint64(len(p)))
-			stats.MemReadLatency.SampleTimeSince(t1)
-		}()
-		n = copy(p, data[off:])
-		if n < len(p) {
-			err = io.ErrUnexpectedEOF
-		}
-		return
-	}
-
-	defer func() {
-		stats.DynamoBytesPerRead.Sample(uint64(len(p)))
-		stats.DynamoReadLatency.SampleTimeSince(t1)
-	}()
-	return dtr.readRange(p, off)
-}
-
-func (dtr *dynamoTableReader) readRange(p []byte, off int64) (n int, err error) {
-	data, err := tryDynamoTableRead(dtr.ddb, dtr.table, dtr.h)
+func (dtra *dynamoTableReaderAt) ReadAtWithStats(p []byte, off int64, stats *Stats) (n int, err error) {
+	data, err := dtra.ddb.ReadTable(dtra.h, stats)
 	d.PanicIfError(err)
-
-	dynamoTableCacheMaybeAdd(dtr.tc, dtr.hash(), data)
-
 	n = copy(p, data[off:])
 	if n < len(p) {
 		err = io.ErrUnexpectedEOF
 	}
 	return
 }
 
-func dynamoTableCacheMaybeGet(tc *sizecache.SizeCache, name addr) (data []byte, present bool) {
-	if tc != nil {
-		if i, present := tc.Get(name); present {
-			return i.([]byte), true
+type ddbTableStore struct {
+	ddb    ddbsvc
+	table  string
+	readRl chan struct{}
+	cache  *sizecache.SizeCache // TODO: merge this with tableCache as part of BUG 3601
+}
+
+func (dts *ddbTableStore) ReadTable(name addr, stats *Stats) (data []byte, err error) {
```

Comment:
`tableMayBeInDynamo()` is provided by `awsLimits`, and chunkMax and the other things in there are subject to change in future patches. As you pointed out, it's conceivable that we might find ourselves in a situation where the current version of the code decides a table "can't" be in Dynamo according to current policy, but a _previous_ version of the code had a different policy. In that case we'd go try to read it from S3 and discover it's not there (because an earlier version of the server had written the table to Dynamo) and so we'd want to fail over to checking in Dynamo in that case.

Given that I kind of expect that situation in the future, and that baking the `awsLimits` into this little object here seemed like it wouldn't make the current code THAT much simpler, I decided to do it this way :-)